File: app/src/main/java/com/ternaryop/photoshelf/parsers/TitleParserConfig.java
package com.ternaryop.photoshelf.parsers;

import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Created by dave on 21/03/15.
 * Access to TitleParser configuration
 */
public interface TitleParserConfig {
    List<TitleParserRegExp> getTitleCleanerList();
    String applyList(List<TitleParserRegExp> titleParserRegExpList, String input);
    Pattern getTitleParserPattern();
    List<LocationPrefix> getLocationPrefixes();
    Map<String, Pattern> getCities();

    class TitleParserRegExp {
        final String pattern;
        final String replacer;

        public TitleParserRegExp(final String pattern, final String replacer) {
            this.pattern = pattern;
            this.replacer = replacer;
        }

        public static String applyList(final List<TitleParserRegExp> titleParserRegExpList, final String input) {
            String result = input;

            for (TitleParserRegExp re : titleParserRegExpList) {
                result = result.replaceAll(re.pattern, re.replacer);
            }
            return result;
        }
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/parsers/TitleParserConfig.kt
package com.ternaryop.photoshelf.parsers

import java.util.regex.Pattern

/**
 * Created by dave on 21/03/15.
 * Access to TitleParser configuration
 */
interface TitleParserConfig {
    val titleCleanerList: List<TitleParserRegExp>
    val titleParserPattern: Pattern
    val locationPrefixes: List<LocationPrefix>
    val cities: Map<String, Pattern>
    fun applyList(titleParserRegExpList: List<TitleParserRegExp>, input: String): String

    class TitleParserRegExp(internal val pattern: String, internal val replacer: String) {
        companion object {

            fun applyList(titleParserRegExpList: List<TitleParserRegExp>, input: String): String {
                var result = input

                for (re in titleParserRegExpList) {
                    result = result.replace(re.pattern.toRegex(), re.replacer)
                }
                return result
            }
        }
    }
}
